### ISSUE TITLE
NCCL backend fails when calling broadcast from different threads

### ISSUE THREAD

**Issue by deepakn94** [pytorch/pytorch#18300]
## 🐛 Bug

When I use the NCCL backend, `torch.distributed.broadcast()` results in a `NCCL error` when called from multiple threads (each thread uses a separate process group). `broadcast()` works fine from multiple threads when using the Gloo backend; NCCL backend works fine in the same environment when using a single thread.

## To Reproduce

Code to reproduce:

```python
import argparse
import os
import threading
import time
import torch
import torch.distributed as dist

NUM_TRIALS = 20

def receive_tensor_helper(tensor, src_rank, group, tag, num_iterations,
                          intra_server_broadcast):
    for i in range(num_iterations):
        if intra_server_broadcast:
            dist.broadcast(tensor=tensor, group=group, src=src_rank)
        else:
            dist.recv(tensor=tensor, src=src_rank, tag=tag)
    print("Done with tensor size %s" % tensor.size())

def send_tensor_helper(tensor, dst_rank, group, tag, num_iterations,
                       intra_server_broadcast):
    for i in range(num_iterations):
        if intra_server_broadcast:
            dist.broadcast(tensor=tensor, group=group, src=1-dst_rank)
        else:
            dist.send(tensor=tensor, dst=dst_rank, tag=tag)
    print("Done with tensor size %s" % tensor.size())

def start_helper_thread(func, args):
    helper_thread = threading.Thread(target=func,
                                     args=tuple(args))
    helper_thread.start()


if __name__ == '__main__':
    parser = argparse.ArgumentParser(
        description='Test lightweight communication library')
    parser.add_argument("--backend", type=str, default='gloo',
                        help="Backend")
    parser.add_argument('-s', "--send", action='store_true',
                        help="Send tensor (if not specified, will receive tensor)")
    parser.add_argument("--master_addr", required=True, type=str,
                        help="IP address of master")
    parser.add_argument("--use_helper_threads", action='store_true',
                        help="Use multiple threads")
    parser.add_argument("--rank", required=True, type=int,
                        help="Rank of current worker")
    parser.add_argument('-p', "--master_port", default=12345,
                        help="Port used to communicate tensors")
    parser.add_argument("--intra_server_broadcast", action='store_true',
                        help="Broadcast within a server")

    args = parser.parse_args()

    num_ranks_in_server = 1
    if args.intra_server_broadcast:
        num_ranks_in_server = 2
    local_rank = args.rank % num_ranks_in_server
    torch.cuda.set_device(local_rank)

    os.environ['MASTER_ADDR'] = args.master_addr
    os.environ['MASTER_PORT'] = str(args.master_port)
    world_size = 2
    dist.init_process_group(args.backend, rank=args.rank, world_size=world_size)

    tensor_sizes = [10, 100, 1000, 10000, 100000, 1000000, 10000000]

    groups = []
    for tag in range(len(tensor_sizes)):
        if args.intra_server_broadcast:
            group = dist.new_group([0, 1])
            groups.append(group)

    for tag, tensor_size in enumerate(tensor_sizes):
        if args.intra_server_broadcast:
            group = groups[tag]
        else:
            group = None

        if args.send:
            if args.intra_server_broadcast:
                tensor = torch.tensor(range(tensor_size), dtype=torch.float32).cuda(
                    local_rank)
            else:
                tensor = torch.tensor(range(tensor_size), dtype=torch.float32).cpu()
            if args.use_helper_threads:
                start_helper_thread(send_tensor_helper,
                                    [tensor, 1-args.rank,
                                     group, tag, NUM_TRIALS,
                                     args.intra_server_broadcast])
            else:
                send_tensor_helper(tensor, 1-args.rank, group, tag,
                                   NUM_TRIALS, args.intra_server_broadcast)
        else:
            if args.intra_server_broadcast:
                tensor = torch.zeros((tensor_size,), dtype=torch.float32).cuda(
                    local_rank)
            else:
                tensor = torch.zeros((tensor_size,), dtype=torch.float32).cpu()
            if args.use_helper_threads:
                start_helper_thread(receive_tensor_helper,
                                    [tensor, 1-args.rank,
                                     group, tag, NUM_TRIALS,
                                     args.intra_server_broadcast])
            else:
                receive_tensor_helper(tensor, 1-args.rank, group, tag,
                                      NUM_TRIALS, args.intra_server_broadcast)
```

Commands to reproduce the error (on the same multi-GPU server):
```bash
python harness.py --master_addr localhost --rank 0 --intra_server_broadcast --backend nccl --use_helper_threads --send
python harness.py --master_addr localhost --rank 1 --intra_server_broadcast --backend nccl --use_helper_threads
```

Error trace:
```bash
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/opt/conda/lib/python3.6/threading.py", line 864, in run
    self._target(*self._args, **self._kwargs)
  File "harness.py", line 15, in receive_tensor_helper
    dist.broadcast(tensor=tensor, group=group, src=src_rank)
  File "/opt/conda/lib/python3.6/site-packages/torch/distributed/distributed_c10d.py", line 741, in broadcast
    work = group.broadcast([tensor], opts)
RuntimeError: NCCL error in: ../torch/lib/c10d/ProcessGroupNCCL.cpp:260, unhandled system error
```

These commands work as expected:
```bash
python harness.py --master_addr localhost --rank 0 --intra_server_broadcast --backend gloo --use_helper_threads --send
python harness.py --master_addr localhost --rank 1 --intra_server_broadcast --backend gloo --use_helper_threads
```
```bash
python harness.py --master_addr localhost --rank 0 --intra_server_broadcast --backend nccl --send
python harness.py --master_addr localhost --rank 1 --intra_server_broadcast --backend nccl
```

## Expected behavior

No runtime error / tensors successfully broadcasted among participating ranks.

## Environment

Please copy and paste the output from our [environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

```
Collecting environment information...
PyTorch version: 1.0.0a0+056cfaf
Is debug build: No
CUDA used to build PyTorch: 10.0.130

OS: Ubuntu 16.04.5 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.11) 5.4.0 20160609
CMake version: version 3.5.1

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.0.130
GPU models and configuration:
GPU 0: Tesla V100-PCIE-16GB
GPU 1: Tesla V100-PCIE-16GB
GPU 2: Tesla V100-PCIE-16GB
GPU 3: Tesla V100-PCIE-16GB

Nvidia driver version: 384.130
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.4.2

Versions of relevant libraries:
[pip] msgpack-numpy==0.4.3.2
[pip] numpy==1.15.4
[pip] torch==1.0.0a0+056cfaf
[pip] torchtext==0.4.0
[pip] torchvision==0.2.1
[conda] blas                      1.0                         mkl
[conda] magma-cuda100             2.4.0                         1    pytorch
[conda] mkl                       2018.0.3                      1
[conda] mkl-include               2019.1                      144
[conda] mkl_fft                   1.0.6            py36h7dd41cf_0
[conda] mkl_random                1.0.1            py36h4414c95_1
[conda] torch                     1.0.0a0+056cfaf           <pip>
[conda] torchtext                 0.4.0                     <pip>
[conda] torchvision               0.2.1                     <pip>
```

**Comment by mrshenli** [comment]
Hi @deepakn94, thanks for reporting the bug. The error is caused by multiple threads accessing the same underlying store. Even though each thread has its own process group (pg), those pgs all [share the same default store](https://github.com/pytorch/pytorch/blob/master/torch/distributed/distributed_c10d.py#L419). Hence, the [set and get in NCCL pg](https://github.com/pytorch/pytorch/blob/master/torch/lib/c10d/ProcessGroupNCCL.cpp#L189-L202) hit race conditions. Holding the [lock](https://github.com/pytorch/pytorch/blob/master/torch/lib/c10d/ProcessGroupNCCL.cpp#L183) until after get/set solves the problem. 

@pietern just want to check with you before I submit the fix. Is multi-thread with the same underlying store an expected scenario? Or do we just update docs to clarify which scenarios are supported?

**Comment by deepakn94** [comment]
Hi @mrshenli, thanks for looking at this!

Just to confirm, the lock should be held till [this line](https://github.com/pytorch/pytorch/blob/master/torch/lib/c10d/ProcessGroupNCCL.cpp#L203), correct? And should not be held for the `memcpy`?

**Comment by mrshenli** [comment]
@deepakn94 yes, and also [this line](https://github.com/pytorch/pytorch/blob/master/torch/lib/c10d/ProcessGroupNCCL.cpp#L194)

**Comment by deepakn94** [comment]
Hi @mrshenli, I tried this fix and am getting a similar error to before,

```bash
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/opt/conda/lib/python3.6/threading.py", line 864, in run
    self._target(*self._args, **self._kwargs)
  File "harness.py", line 24, in send_tensor_helper
    dist.broadcast(tensor=tensor, group=group, src=1-dst_rank)
  File "/opt/conda/lib/python3.6/site-packages/torch/distributed/distributed_c10d.py", line 743, in broadcast
    work = group.broadcast([tensor], opts)
RuntimeError: NCCL error in: ../torch/lib/c10d/ProcessGroupNCCL.cpp:260, unhandled cuda error
```

This is the change I made,
```bash
diff --git a/torch/lib/c10d/ProcessGroupNCCL.cpp b/torch/lib/c10d/ProcessGroupNCCL.cpp
index 538c45b..b029643 100644
--- a/torch/lib/c10d/ProcessGroupNCCL.cpp
+++ b/torch/lib/c10d/ProcessGroupNCCL.cpp
@@ -180,8 +180,6 @@ void ProcessGroupNCCL::broadcastUniqueNCCLID(ncclUniqueId* ncclID) {
   auto groupPgId = groupName_ + "_" + processGroupID_;
   const auto uniqueNCCLIDCnt = ++pgUniqueNCCLIDCnt_[groupPgID_];

-  lock.unlock();
-
   std::string storeKey =
       processGroupID_ + "_" + std::to_string(uniqueNCCLIDCnt);

@@ -191,6 +189,7 @@ void ProcessGroupNCCL::broadcastUniqueNCCLID(ncclUniqueId* ncclID) {
         reinterpret_cast<uint8_t*>(ncclID),
         reinterpret_cast<uint8_t*>(ncclID) + NCCL_UNIQUE_ID_BYTES);
     store_->set(storeKey, ncclIDVal);
+    lock.unlock();
     // Other ranks get to the store
   } else {
     auto ncclIDVal = store_->get(storeKey);
@@ -202,6 +201,7 @@ void ProcessGroupNCCL::broadcastUniqueNCCLID(ncclUniqueId* ncclID) {
     }
     // Now put the data back to the input pointer
     memcpy(ncclID, ncclIDVal.data(), NCCL_UNIQUE_ID_BYTES);
+    lock.unlock();
   }
 }
```

I also tried moving the `lock.unlock()` before the `memcpy` with no success.

**Comment by mrshenli** [comment]
@deepakn94 last time when I debug, I reduced `NUM_TRIALS` to 2.  After setting `NUM_TRIALS` back to 20, I cannot reproduce the original error but I do see the receiver hangs on input tensor size `1000000` and `10000000`. I will look into this.

**Comment by pietern** [comment]
Because you create the new groups through `new_group` they will all use a different prefix when using the store and shouldn't interfere at that level. It is possible there is some underlying thing in NCCL that means they do, but you'd need to go through the NCCL source code to confirm this.

Can you run with NCCL_DEBUG=INFO and see what you get? Perhaps it includes some hint as to what's going on here...

**Comment by deepakn94** [comment]
Sorry, I've been busy with a deadline! Will get back to you on this by Thursday or Friday this week!

**Comment by deepakn94** [comment]
This is the output with `NCCL_DEBUG=INFO` (and then it hangs),

```bash
v08:1707:1718 [0] NCCL INFO NET : Using interface eth0:10.0.0.29<0>
v08:1707:1718 [0] NCCL INFO NET/IB : Using interface eth0 for sideband communication
v08:1707:1718 [0] NCCL INFO Using internal Network Socket
v08:1707:1718 [0] NCCL INFO NET : Using interface eth0:10.0.0.29<0>
v08:1707:1718 [0] NCCL INFO NET : Using interface lxdbr0:fe80::6cde:27ff:fe94:d560%lxdbr0<0>
v08:1707:1718 [0] NCCL INFO NET/Socket : 2 interfaces found
NCCL version 2.3.7+cuda10.0
v08:1707:1718 [0] NCCL INFO rank 0 nranks 2
v08:1707:1719 [0] NCCL INFO rank 0 nranks 2
v08:1707:1720 [0] NCCL INFO rank 0 nranks 2
v08:1707:1722 [0] NCCL INFO rank 0 nranks 2
v08:1707:1721 [0] NCCL INFO rank 0 nranks 2
v08:1707:1728 [0] NCCL INFO rank 0 nranks 2
v08:1707:1730 [0] NCCL INFO comm 0x7f4bec028280 rank 0 nranks 2
v08:1707:1735 [0] NCCL INFO comm 0x7f4bdc0014f0 rank 0 nranks 2
v08:1707:1732 [0] NCCL INFO comm 0x7f4be4000de0 rank 0 nranks 2
v08:1707:1733 [0] NCCL INFO comm 0x7f4be0000de0 rank 0 nranks 2
v08:1707:1734 [0] NCCL INFO comm 0x7f4be8000de0 rank 0 nranks 2
v08:1707:1731 [0] NCCL INFO comm 0x7f4bf0000de0 rank 0 nranks 2
v08:1707:1736 [0] NCCL INFO rank 0 nranks 2
v08:1707:1738 [0] NCCL INFO comm 0x7f4bc8000de0 rank 0 nranks 2
```

Without `--use_helper_threads`, it doesn't hang and the output looks like this,
```bash
v08:1797:1797 [0] NCCL INFO NET : Using interface eth0:10.0.0.29<0>
v08:1797:1797 [0] NCCL INFO NET/IB : Using interface eth0 for sideband communication
v08:1797:1797 [0] NCCL INFO Using internal Network Socket
v08:1797:1797 [0] NCCL INFO NET : Using interface eth0:10.0.0.29<0>
v08:1797:1797 [0] NCCL INFO NET : Using interface lxdbr0:fe80::6cde:27ff:fe94:d560%lxdbr0<0>
v08:1797:1797 [0] NCCL INFO NET/Socket : 2 interfaces found
NCCL version 2.3.7+cuda10.0
v08:1797:1797 [0] NCCL INFO rank 0 nranks 2
v08:1798:1798 [1] NCCL INFO NET : Using interface eth0:10.0.0.29<0>
v08:1798:1798 [1] NCCL INFO NET/IB : Using interface eth0 for sideband communication
v08:1798:1798 [1] NCCL INFO Using internal Network Socket
v08:1798:1798 [1] NCCL INFO rank 1 nranks 2
v08:1797:1809 [0] NCCL INFO comm 0x7fd2c0028910 rank 0 nranks 2
v08:1798:1810 [1] NCCL INFO comm 0x7f3194028910 rank 1 nranks 2
v08:1798:1810 [1] NCCL INFO NET : Using interface eth0:10.0.0.29<0>
v08:1798:1810 [1] NCCL INFO NET : Using interface lxdbr0:fe80::6cde:27ff:fe94:d560%lxdbr0<0>
v08:1798:1810 [1] NCCL INFO NET/Socket : 2 interfaces found
v08:1798:1810 [1] NCCL INFO Could not find real path of /sys/class/net/lxdbr0/device
v08:1798:1810 [1] NCCL INFO CUDA Dev 1, IP Interfaces : eth0(PXB) lxdbr0(SOC)
v08:1797:1809 [0] NCCL INFO Could not find real path of /sys/class/net/lxdbr0/device
v08:1797:1809 [0] NCCL INFO CUDA Dev 0, IP Interfaces : eth0(PXB) lxdbr0(SOC)
v08:1797:1809 [0] NCCL INFO Using 256 threads
v08:1797:1809 [0] NCCL INFO Min Comp Cap 7
v08:1797:1809 [0] NCCL INFO Ring 00 :    0   1
v08:1797:1809 [0] NCCL INFO Ring 00 : 0[0] -> 1[1] via direct shared memory
v08:1798:1810 [1] NCCL INFO Ring 00 : 1[1] -> 0[0] via direct shared memory
v08:1798:1810 [1] NCCL INFO comm 0x7f3194028910 rank 1 nranks 2 - COMPLETE
v08:1797:1809 [0] NCCL INFO comm 0x7fd2c0028910 rank 0 nranks 2 - COMPLETE
v08:1797:1797 [0] NCCL INFO Launch mode Parallel
v08:1797:1797 [0] NCCL INFO rank 0 nranks 2
v08:1798:1798 [1] NCCL INFO rank 1 nranks 2
v08:1797:1812 [0] NCCL INFO comm 0x7fd2c8001610 rank 0 nranks 2
v08:1798:1813 [1] NCCL INFO comm 0x7f319402d1c0 rank 1 nranks 2
v08:1798:1813 [1] NCCL INFO Could not find real path of /sys/class/net/lxdbr0/device
v08:1798:1813 [1] NCCL INFO CUDA Dev 1, IP Interfaces : eth0(PXB) lxdbr0(SOC)
v08:1797:1812 [0] NCCL INFO Could not find real path of /sys/class/net/lxdbr0/device
v08:1797:1812 [0] NCCL INFO CUDA Dev 0, IP Interfaces : eth0(PXB) lxdbr0(SOC)
v08:1797:1812 [0] NCCL INFO Using 256 threads
v08:1797:1812 [0] NCCL INFO Min Comp Cap 7
v08:1797:1812 [0] NCCL INFO Ring 00 :    0   1
```

**Comment by pietern** [comment]
I notice you start those threads but don't join them explicitly. They are likely being joined at process termination, but then other stuff may have already been destructed. Can you try joining them inline (right after calling `start`) and report back what happens?

**Comment by deepakn94** [comment]
If I join inline, there's no error (but no concurrency, so this is similar to not passing in `--helper_threads` at all).

If I join after starting all the helper threads (using the code below), I get the same error as before.

```python
# helper_threads is a list with all created helper_thread handles.

for helper_thread in helper_threads:
    helper_thread.join()
```

**Comment by pietern** [comment]
Ran a few tests to confirm the results you are seeing. I'm thinking there is some sort of global state that is mutated by NCCL that prevents concurrent initialization of communicators. If you strictly sequence the first collective against the same set of devices (`dist.broadcast(torch.zeros(1).cuda(local_rank), group=group,src=0)` before launching the thread) the example works without issues.

**Comment by deepakn94** [comment]
@pietern, so if I do `dist.broadcast(torch.zeros(1).cuda(local_rank), group=group,src=0)` (and the corresponding send on the sending process) for each group in the main thread, things should work even after I launch the threads (and do broadcasts on them)?

**Comment by pietern** [comment]
@deepakn94 Yes, that's what I confirmed to work based on your repro script. What happens then is that the lazy initialization is done sequentially on the main thread. Once that is done, you're free to use the communicators from multiple threads. It seems that initialization doesn't like arbitrary ordering. See the linked issue at nvidia/nccl#239 for more discussion.

**Comment by deepakn94** [comment]
Got it! Will try in my larger application and report back!

**Comment by pietern** [comment]
The discussion in the linked issue includes explanation of the underlying issue.

TL;DR, it's not possible to do concurrent and out of order initialization of NCCL communicators. The hang you're seeing is unavoidable and it requires you to initialize the communicators before using them concurrently. Also, the fact that you CAN use multiple NCCL communicators concurrently in the first place, is a function of the number of NCCL kernels that can run at the same time, and therefore a function of the CUDA runtime, the hardware, etc. I think we'll have to discourage this pattern in the future...

**Comment by pietern** [comment]
@deepakn94 Can you elaborate on what you're using multiple process groups for? Perhaps there is some other way we can achieve the same result.

**Comment by deepakn94** [comment]
At a high level, I'm trying to do peer-to-peer communication of GPU tensors using the NCCL backend (for example, I might send some tensors between GPUs 1 and 2, and some other tensors between GPUs 1 and 3). I want to parallelize this as much as possible, which is the reason for communicating different tensors via broadcast in different helper threads. Ideally, I would want to use something similar to `send` and `recv` to do this peer-to-peer communication, but this doesn't seem possible with the NCCL backend, so am using `broadcast` as a workaround.

**Comment by gchanan** [comment]
@mrshenli any update on this?

**Comment by mrshenli** [comment]
Hey @deepakn94, does @pietern's suggestion work for you?

**Comment by deepakn94** [comment]
It worked on that simple use case, but I'm running into issues on a more complex use case. Will try to come up with a minimal working example that you can play with.

**Comment by Daulbaev** [comment]
@deepakn94 Hi! Is there any progress? 
I also need point-to-point communication between GPUs in PyTorch and want to emulate this feature using broadcast with groups.

**Comment by deepakn94** [comment]
Sorry, I've been busy with a deadline next week. Hopefully will be able to put something reproducible after that.

**Comment by aoyulong** [comment]
Hi, I have the same problem and also want to use broadcast to do send/recv. Is there any way to make sure the communicators to be initialized in order in the main thread before launch the other helper thread? @deepakn94 @pietern

**Comment by mrshenli** [comment]
Hey @aoyulong, NCCL in general is not thread safe: https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/usage/threadsafety.html

We were also told that using multiple NCCL communicators per device could also result in NCCL hang. 

In your application, is it possible to do all NCCL broadcast in one thread?

**Comment by aoyulong** [comment]
@mrshenli In my application, each thread use a different group and so will use different communicators. I think this satisfy the nccl thread safe requirement ` Multiple threads should use separate communicator objects`. Besides, I overlap the computation and communication in the pipeline parallel: 1) the main thread will get data from the recv queue, feed the model, store output to send queue; 2) one recv thread will recv data from its previous stage and store the data into the recv queue; 2) another send thread will get data from the send queue and send to its next stage.  Is there any good way to do this in PyTorch?

**Comment by rohan-varma** [comment]
AFAIK, separate process groups across different threads aren't supported with the NCCL backend, so applications are limited to single-threading. @mrshenli given this can we close this issue out?
